core/storage/email/gae_models_test.py

```diff
@@ -128,9 +149,101 @@ def test_get_by_hash_behavior_with_sent_datetime_lower_bound(self):
+            '_generate_unique_reply_to_id',
+            types.MethodType(
+                _generate_unique_id_for_tests,
+                email_models.GeneralFeedbackEmailReplyToIdModel))
+
+    def test_id_generation_works_correctly(self):
+        model = email_models.GeneralFeedbackEmailReplyToIdModel
+
+        created = model.create('user_id', 'thread_id')
+        self.assertEqual(created.id, 'user_id.thread_id')
+
+        created = model.create('other_user_id', 'other_thread_id')
+        self.assertNotEqual(created.id, 'user_id.thread_id')
+
+    def test_generated_reply_to_id_is_unique(self):
+        model = email_models.GeneralFeedbackEmailReplyToIdModel
+
+        all_ids = [
+            model.create('user_id', 'thread_id').reply_to_id
+            for _ in xrange(100)]
+        set_ids = set(all_ids) #Set contains unique values
+        self.assertEqual(len(all_ids), len(set_ids))
+
+    def test_create_raises_when_duplicate_ids(self):
+        with self.generate_constant_id_ctx, self.assertRaises(Exception):
+            created1 = email_models.GeneralFeedbackEmailReplyToIdModel.create(
+                'user_id', 'thread_id')
+            created1.put()
+            created2 = email_models.GeneralFeedbackEmailReplyToIdModel.create(
+                'user_id', 'thread_id')
+            created2.put()
+
+    def test_get_by_reply_to_id_works_correctly(self):
+        model = email_models.GeneralFeedbackEmailReplyToIdModel
+
+        created = model.create('user_id', 'thread_id')
+        created.put()
+
+        result = model.get_by_reply_to_id(created.reply_to_id)
+
+        self.assertNotEqual(result, None) #There should be a result
+        self.assertEqual(result.reply_to_id, created.reply_to_id)
+
+        result = model.get_by_reply_to_id('non_existent_reply_to_id')
+
+        self.assertEqual(result, None)
+
+    def test_get_works_correctly(self):
+        created = email_models.GeneralFeedbackEmailReplyToIdModel.create(
+            'user_id', 'thread_id')
+        created.put()
+
+        result = email_models.GeneralFeedbackEmailReplyToIdModel.get(
+            'user_id', 'thread_id', strict=False)
+
+        self.assertNotEqual(result, None)
+        self.assertIn('user_id.thread_id', str(result.key))
+
+        result = email_models.GeneralFeedbackEmailReplyToIdModel.get(
+            'bad_user_id', 'bad_thread_id', strict=False) #Should not throw
+        self.assertEqual(result, None)
+
+        with self.assertRaises(Exception):
+            result = email_models.GeneralFeedbackEmailReplyToIdModel.get(
+                'bad_user_id', 'bad_thread_id') #Should throw
+
+    def test_get_multi_by_user_ids_works_correctly(self):
+        model = email_models.GeneralFeedbackEmailReplyToIdModel
+
+        model.create('user_id_1', 'thread_id')
+        model.create('user_id_2', 'thread_id')
+
+        result = model.get_multi_by_user_ids(
+            ['user_id_1', 'user_id_2'], 'thread_id')
```

Comment:
Is it worth also testing:
- One user ID (to mimic the non-multi case)
- No user IDs (pathological case)
- One of the user IDs in the list not corresponding to a real thread, but the other does

Any other worthwhile cases?

Reply:
All three are done, and it allowed me to find out that I should have been more thorough. Better now I think.

Reply:
What do you think @BenHenning ?

Reply:
Bumping up the thread since it got tagged stale, I'm still waiting for review @BenHenning

Reply:
I'm not giving up ! @BenHenning, could you review my last changes ?